crates/esplora/src/async_ext.rs

```diff
@@ -230,176 +237,235 @@ async fn chain_update<A: Anchor>(
+                async move {
+                    if tx_already_exists {
+                        client
+                            .get_tx_status(&txid)
+                            .await
+                            .map(|s| (txid, EsploraResp::TxStatus(s)))
+                    } else {
+                        client
+                            .get_tx_info(&txid)
+                            .await
+                            .map(|t| (txid, EsploraResp::Tx(t)))
+                    }
+                }
             })
             .collect::<FuturesOrdered<_>>();
 
         if handles.is_empty() {
             break;
         }
 
-        for (txid, status) in handles.try_collect::<Vec<(Txid, TxStatus)>>().await? {
-            if let Some(anchor) = anchor_from_status(&status) {
-                let _ = graph.insert_anchor(txid, anchor);
+        for (txid, resp) in handles.try_collect::<Vec<_>>().await? {
+            match resp {
+                EsploraResp::TxStatus(status) => {
+                    insert_anchor_from_status(&mut tx_graph, txid, status);
+                }
+                EsploraResp::Tx(Some(tx_info)) => {
+                    let _ = tx_graph.insert_tx(tx_info.to_tx());
+                    insert_anchor_from_status(&mut tx_graph, txid, tx_info.status);
+                    insert_prevouts(&mut tx_graph, tx_info.vin);
+                }
+                _ => continue,
             }
         }
     }
+    Ok(tx_graph)
+}
 
-    for op in outpoints.into_iter() {
-        if graph.get_tx(op.txid).is_none() {
-            if let Some(tx) = client.get_tx(&op.txid).await? {
-                let _ = graph.insert_tx(tx);
-            }
-            let status = client.get_tx_status(&op.txid).await?;
-            if let Some(anchor) = anchor_from_status(&status) {
-                let _ = graph.insert_anchor(op.txid, anchor);
-            }
+/// Fetch transactions and [`ConfirmationBlockTime`]s that contain and spend the provided
+/// `outpoints`.
+///
+/// `parallel_requests` specifies the maximum number of HTTP requests to make in parallel.
+///
+/// Refer to [crate-level docs](crate) for more.
+async fn fetch_txs_with_outpoints<I: IntoIterator<Item = OutPoint> + Send>(
+    client: &esplora_client::AsyncClient,
+    outpoints: I,
+    parallel_requests: usize,
+) -> Result<TxGraph<ConfirmationBlockTime>, Error>
+where
+    I::IntoIter: Send,
+{
+    let outpoints = outpoints.into_iter().collect::<Vec<_>>();
+
+    // make sure txs exists in graph and tx statuses are updated
+    // TODO: We should maintain a tx cache (like we do with Electrum).
+    let mut tx_graph = fetch_txs_with_txids(
+        client,
+        // NOTE: @evanlinjin We need to collect here otherwise we run into lifetime issues, but I
+        // do not know why. If someone can figure this out, that will be amazing.
+        outpoints.iter().map(|op| op.txid).collect::<Vec<_>>(),
```

Comment:
not sure if this is what you want, but it seems to compile

```suggestion
        outpoints.iter().copied().map(|op| op.txid),
```

Reply:
Thank you @ValuedMammal, that seems to work. I'm still quite confused why your solution works and why `outpoints.iter().map(|op| op.txid)` does not. If someone can explain it to me, that will be wonderful.

Reply:
I found this?

https://users.rust-lang.org/t/implementation-of-send-is-not-general-enough-but-cannot-make-it-more-general/115087